### PR TITLE
UserIntent classes prototype.

### PR DESCRIPTION
**Pull request by epawlowska**
Includes:
- creating sample UserRuleIntent
- UserParserIntent, UserConstraintIntent, UserRuleIntent

### CODE REVIEW

File: whylog/teacher/mockoutputs.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from whylog.teacher.userintent import UserParserIntent, UserConstraintIntent, UserRuleIntent
```
Comment (ppolewicz) [inline whylog/teacher/mockoutputs.py]: sort imports

File: whylog/teacher/__init__.py [hunk 2]
```diff
@@ -52,4 +56,4 @@ def set_causes_relation(self, relation):
         pass
 
     def save(self):
-        pass
+        self.config.add_rule(create_sample_rule())
```
Comment (ppolewicz) [inline whylog/teacher/__init__.py:59]: This line suggests that the caller of config does not need to call `config.save()` after finishing modifying configuration, which will require O(N) rewrites of the config to storage (file/database/etc). Maybe it should be a context manager (which translates to open file for writing / transaction) or something?

File: whylog/teacher/mockoutputs.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    content1 = "2015-12-03 12:08:09 Connection error occurred on alfa36. Host name: 2"
+    content2 = "2015-12-03 12:10:10 Data migration from alfa36 to alfa21 failed. Host name: 2"
+    content3 = "2015-12-03 12:11:00 Data is missing at alfa21. Loss = 567.02 GB. Host name: 101"
+    """
+
+    regex1 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Connection error occurred on (.*). Host name: (.*)$"
+    regex2 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Data migration from (.*) to (.*) failed. Host name: (.*)$"
+    regex3 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Data is missing at (.*). Loss = (.*) GB. Host name; (.*)$"
+
+    parser1 = UserParserIntent("hydra", regex1, [1], dict([(1, to_date)]))
```
Comment (ppolewicz) [inline whylog/teacher/mockoutputs.py]: `dict([(1, to_date)])` -> `{1: to_date}`. The same applies for 2 following lines.

File: whylog/teacher/mockoutputs.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    # default: TO_STRING = 1
+    """
+    content1 = "2015-12-03 12:08:09 Connection error occurred on alfa36. Host name: 2"
+    content2 = "2015-12-03 12:10:10 Data migration from alfa36 to alfa21 failed. Host name: 2"
+    content3 = "2015-12-03 12:11:00 Data is missing at alfa21. Loss = 567.02 GB. Host name: 101"
+    """
+
+    regex1 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Connection error occurred on (.*). Host name: (.*)$"
+    regex2 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Data migration from (.*) to (.*) failed. Host name: (.*)$"
+    regex3 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Data is missing at (.*). Loss = (.*) GB. Host name; (.*)$"
```
Comment (ppolewicz) [inline whylog/teacher/mockoutputs.py]: `Host name;` -> `Host name:`

File: whylog/teacher/mockoutputs.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    to_date = 0
+    to_float = 3
+    # default: TO_STRING = 1
+    """
+    content1 = "2015-12-03 12:08:09 Connection error occurred on alfa36. Host name: 2"
+    content2 = "2015-12-03 12:10:10 Data migration from alfa36 to alfa21 failed. Host name: 2"
+    content3 = "2015-12-03 12:11:00 Data is missing at alfa21. Loss = 567.02 GB. Host name: 101"
+    """
+
+    regex1 = "^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) Connection error occurred on (.*). Host name: (.*)$"
```
Comment (ppolewicz) [inline whylog/teacher/mockoutputs.py]: escape `.` in regexes properly or they will match any character

File: whylog/teacher/userintent.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class UserParserIntent(object):
```
Comment (ppolewicz) [inline whylog/teacher/userintent.py]: rename module to `user_intent.py`

File: whylog/teacher/userintent.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    def __init__(self, log_type_name, regex, primary_key_groups, data_conversions=None):
+        self.log_type_name = log_type_name
+        self.regex = regex
+        self.primary_key_groups = primary_key_groups
+        self.data_conversions = data_conversions
```
Comment (ppolewicz) [inline whylog/teacher/userintent.py]:
``` python
self.data_conversions = data_conversions or {}
```

File: whylog/teacher/userintent.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class UserConstraintIntent(object):
+    def __init__(self, constr_type, groups, params=None):
+        self.type = constr_type
+        self.groups = groups
+        self.params = params
```
Comment (ppolewicz) [inline whylog/teacher/userintent.py]:
``` python
self.params = params or {}
```

this way every user of `self.params` will not need to check for `is None`

File: whylog/teacher/mockoutputs.py (hunk 1 above)
Comment (ppolewicz) [inline whylog/teacher/mockoutputs.py]: rename module to `mock_outputs.py` to follow standard

File: whylog/teacher/userintent.py (hunk 7 above)
Comment (epawlowska) [inline whylog/teacher/userintent.py]: This 'or' is cool! Normally I would write sth like "{} if data_conversions is None else data_conversions". Thanks!

**Pull request merged**